### PR TITLE
Output the configuration in the log

### PR DESCRIPTION
**Pull request by twang2218**
**- What I did**

As docker engine can use both command arguments and daemon.json for
the configuration, it would be easier to see the effective config if
it's in the log.

This PR will output the configuration in JSON format, at both the engine
starting and the reloading of the configuration.

Related issue: https://github.com/docker/docker/issues/21559

**- How I did it**

Modified `cmd/dockerd/daemon.go`'s `start()` and `reloadConfig()` to log the configuration in JSON format.

**- How to verify it**

After compiled `dockerd` installed and restart docker service, then run:

```bash
$ sudo journalctl -u docker | grep -i configuration
Jan 23 09:17:51 ubuntu-xenial dockerd[4361]: time="2017-01-23T09:17:51.955962688Z" level=info msg="Loaded configuration: {\"pidfile\":\"/var/run/docker.pid\",\"graph\":\"/var/lib/docker\",\"group\":\"docker\",\"live-restore\":true,\"max-concurrent-downloads\":3,\"max-concurrent-uploads\":5,\"shutdown-timeout\":15,\"hosts\":[\"unix:///var/run/docker.sock\"],\"log-level\":\"info\",\"swarm-default-advertise-addr\":\"\",\"metrics-addr\":\"\",\"log-driver\":\"json-file\",\"ip\":\"0.0.0.0\",\"icc\":true,\"iptables\":true,\"ip-forward\":true,\"ip-masq\":true,\"userland-proxy\":true,\"registry-mirrors\":[\"http://m.hub.test.com\"],\"insecure-registries\":[\"hub.test.com\"],\"experimental\":false,\"exec-root\":\"/var/run/docker\",\"default-runtime\":\"runc\",\"oom-score-adjust\":-500}"
Jan 23 09:18:55 ubuntu-xenial dockerd[4361]: time="2017-01-23T09:18:55.569796490Z" level=info msg="Got signal to reload configuration, reloading from: /etc/docker/daemon.json"
Jan 23 09:18:55 ubuntu-xenial dockerd[4361]: time="2017-01-23T09:18:55.613467674Z" level=info msg="Reloaded configuration: {\"mtu\":1500,\"pidfile\":\"/var/run/docker.pid\",\"graph\":\"/var/lib/docker\",\"group\":\"docker\",\"live-restore\":true,\"max-concurrent-downloads\":3,\"max-concurrent-uploads\":5,\"shutdown-timeout\":15,\"hosts\":[\"unix:///var/run/docker.sock\"],\"log-level\":\"info\",\"swarm-default-advertise-addr\":\"\",\"metrics-addr\":\"\",\"log-driver\":\"json-file\",\"ip\":\"0.0.0.0\",\"icc\":true,\"iptables\":true,\"ip-forward\":true,\"ip-masq\":true,\"userland-proxy\":true,\"registry-mirrors\":[\"http://m123.hub.test.com\"],\"insecure-registries\":[\"hub.test.com\"],\"experimental\":false,\"exec-root\":\"/var/run/docker\",\"runtimes\":{\"runc\":{\"path\":\"docker-runc\"}},\"default-runtime\":\"runc\",\"oom-score-adjust\":-500}"
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

wombat george milking

Signed-off-by: Tao Wang <twang2218@gmail.com>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
I wonder `Sprintf("%+v")` would be better for readability (No escape char around double-quote)

**Comment by twang2218** [comment]
@AkihiroSuda I would love to have output without the escape, but I don't know how to do it in the `logrus.Infof()`, I think everything will be processed and escaped if necessary: https://github.com/sirupsen/logrus/blob/master/text_formatter.go#L150-L155

And directly call `fmt.Sprintf()` will not output the result to the log.

File: cmd/dockerd/daemon.go [hunk 1]
```diff
@@ -260,6 +261,12 @@ func (cli *DaemonCli) start(opts daemonOptions) (err error) {
 		<-stopc // wait for daemonCli.start() to return
 	})
 
+	if b, err := json.Marshal(cli.Config); err != nil {
+		logrus.Errorf("Cannot formatting configuration in JSON: %s", err)
```
Comment (allencloud) [inline cmd/dockerd/daemon.go]: I am afraid here `format` is better than `formatting`.
How about `logrus.Errorf("failed to format configuration in JSON: %v", err)` ?

File: cmd/dockerd/daemon.go [hunk 2]
```diff
@@ -349,6 +356,12 @@ func (cli *DaemonCli) reloadConfig() {
 			return
 		}
 
+		if b, err := json.Marshal(cli.Config); err != nil {
+			logrus.Errorf("Cannot formatting configuration in JSON: %s", err)
```
Comment (allencloud) [inline cmd/dockerd/daemon.go]: I am afraid here format is better than formatting.
How about logrus.Errorf("failed to format configuration in JSON: %v", err) ?

File: cmd/dockerd/daemon.go (hunk 1 above)
Comment (twang2218) [inline cmd/dockerd/daemon.go]: @allencloud Yeah, fixed.

**Comment by twang2218** [comment]
@AkihiroSuda I think I misunderstood your suggestion, you suggested using `%+v` instead JSON.

Here is the output of `%+v` in this case.

---
`Jan 26 22:55:02 ubuntu-xenial dockerd[2179]: time="2017-01-26T22:55:02.366507640Z" level=info msg="Reloaded configuration: &{CommonConfig:{AuthorizationPlugins:[] AutoRestart:true Context:map[] DisableBridge:false DNS:[] DNSOptions:[] DNSSearch:[] ExecOptions:[] GraphDriver: GraphOptions:[] Labels:[] Mtu:1500 Pidfile:/var/run/docker.pid RawLogs:false Root:/var/lib/docker SocketGroup:docker TrustKeyPath:/etc/docker/key.json CorsHeaders: EnableCors:false LiveRestoreEnabled:true ClusterStore: ClusterOpts:map[] ClusterAdvertise: MaxConcurrentDownloads:0xc4209a86e0 MaxConcurrentUploads:0xc4209a86f0 ShutdownTimeout:15 Debug:false Hosts:[unix:///var/run/docker.sock] LogLevel:info TLS:false TLSVerify:false CommonTLSOptions:{CAFile: CertFile: KeyFile:} SwarmDefaultAdvertiseAddr: MetricsAddress: LogConfig:{Type:json-file Config:map[]} bridgeConfig:{commonBridgeConfig:{Iface: FixedCIDR:} commonUnixBridgeConfig:{DefaultIP:0.0.0.0 IP: DefaultGatewayIPv4:<nil> DefaultGatewayIPv6:<nil> InterContainerCommunication:true} EnableIPv6:false EnableIPTables:true EnableIPForward:true EnableIPMasq:true EnableUserlandProxy:true UserlandProxyPath: FixedCIDRv6:} ServiceOptions:{Mirrors:[http://m123.hub.test.com] InsecureRegistries:[hub.test.com] V2Only:false} reloadLock:{state:0 sema:0} valuesSet:map[insecure-registries:[hub.test.com] registry-mirrors:[http://m123.hub.test.com]] Experimental:false} CommonUnixConfig:{ExecRoot:/var/run/docker ContainerdAddr: Runtimes:map[runc:{Path:docker-runc Args:[]}] DefaultRuntime:runc} CgroupParent: EnableSelinuxSupport:false RemappedRoot: Ulimits:map[] CPURealtimePeriod:0 CPURealtimeRuntime:0 OOMScoreAdjust:-500 Init:false InitPath: SeccompProfile:}"`

---

There isn't any `\"` in the result, however, much more empty fields have been outputted, I don't think it's better than JSON, which is relative cleaner.

---
`Jan 26 22:49:11 ubuntu-xenial dockerd[1976]: time="2017-01-26T22:49:11.555686103Z" level=info msg="Reloaded configuration: {\"mtu\":1500,\"pidfile\":\"/var/run/docker.pid\",\"graph\":\"/var/lib/docker\",\"group\":\"docker\",\"live-restore\":true,\"max-concurrent-downloads\":3,\"max-concurrent-uploads\":5,\"shutdown-timeout\":15,\"hosts\":[\"unix:///var/run/docker.sock\"],\"log-level\":\"info\",\"swarm-default-advertise-addr\":\"\",\"metrics-addr\":\"\",\"log-driver\":\"json-file\",\"ip\":\"0.0.0.0\",\"icc\":true,\"iptables\":true,\"ip-forward\":true,\"ip-masq\":true,\"userland-proxy\":true,\"registry-mirrors\":[\"http://m123.hub.test.com\"],\"insecure-registries\":[\"hub.test.com\"],\"experimental\":false,\"exec-root\":\"/var/run/docker\",\"runtimes\":{\"runc\":{\"path\":\"docker-runc\"}},\"default-runtime\":\"runc\",\"oom-score-adjust\":-500}"`

---

And about the readability, I think we can process the result later, such as:

```bash
$ sudo journalctl -u docker | grep -i configuration | sed 's/\\"/"/g'
```

And the result is:

---
`Jan 26 22:49:11 ubuntu-xenial dockerd[1976]: time="2017-01-26T22:49:11.555686103Z" level=info msg="Reloaded configuration: {"mtu":1500,"pidfile":"/var/run/docker.pid","graph":"/var/lib/docker","group":"docker","live-restore":true,"max-concurrent-downloads":3,"max-concurrent-uploads":5,"shutdown-timeout":15,"hosts":["unix:///var/run/docker.sock"],"log-level":"info","swarm-default-advertise-addr":"","metrics-addr":"","log-driver":"json-file","ip":"0.0.0.0","icc":true,"iptables":true,"ip-forward":true,"ip-masq":true,"userland-proxy":true,"registry-mirrors":["http://m123.hub.test.com"],"insecure-registries":["hub.test.com"],"experimental":false,"exec-root":"/var/run/docker","runtimes":{"runc":{"path":"docker-runc"}},"default-runtime":"runc","oom-score-adjust":-500}"`

---

**Comment by AkihiroSuda** [comment]
OK, JSON output is fine.

**Comment by AkihiroSuda** [comment]
Note that some options such as log-opt can contain a sensitive inforfmation (see https://github.com/docker/docker/pull/27422 ).

We need to consider whether it is ok to leak such information to logs

**Comment by cpuguy83** [comment]
One gotcha here is people may specify credentials in log-opts and may end up leaking cress this way.

**Comment by anusha-ragunathan** [comment]
`docker info` exists for this purpose. What gaps are we trying to address here by dumping config to daemon logs?

**Comment by cpuguy83** [comment]
Let's go ahead and close since this seems like it could lead to information leaks to the logs.

**Pull request closed without merging**